### ISSUE TITLE
[discussion] [python] Generic way for generating module-wrappers for stateless parametrized functions

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#51782]
A lot of functionality like activations is implemented as stateless functions with some scalar parameters. Corresponding modules usually just hold some of these parameters and pass them to the function during forward. This requires to write boiler-plate code for these modules.

An alternative way could be generic standard wrapper generator (at least for python) like `modularize` function below. Then PyTorch could have `nn.Hardtanh = modularize(F.hardtanh)`.

Typing information could also be generated dynamically from the corresponding typing signatures from inner functions.

```
import torch

class Modularize(torch.nn.Module):
    def __init__(self, func, **kwargs):
        super().__init__()
        self.__name__ = func.__name__
        self.forward = lambda *a, **kw: func(*a, **kw, **kwargs)

def modularize(func, __name__ = None, **kwargs):
    return type(__name__ or 'Modularize_' + func.__name__, (torch.nn.Module, ), dict(forward = lambda self, *a, **kw: func(*a, **kw, **kwargs)))

if __name__ == '__main__':
    def add(x, y):
        return x + y

    adder1_instance = Modularize(add, y = 1)
    adder_module = modularize(add, y = 2)
    adder2_instance = adder_module()

    x = torch.zeros(4)
    print(type(adder1_instance), adder1_instance(x))
    print(type(adder2_instance), adder2_instance(x))
```

**Comment by albanD** [comment]
Would you have time to write a proposal for the full API for this?
In particular which arguments from the functional version are expected at Module construction vs at Module call time ?

**Comment by vadimkantorov** [comment]
From above:
```python
def modularize(func, __name__ = None):
    return type(__name__ or 'Modularize_' + func.__name__, (torch.nn.Module, ), dict(__init__ = lambda self, **kwargs: torch.nn.Module.__init__(self) or setattr(self, 'kwargs', kwargs), forward = lambda self, *a, **kw: func(*a, **dict(self.kwargs, **kw))))
```
can be a quick draft

I guess a big question would be to ensure that this traces/scripts okay. And hopefully this module class and corresponding instances should also pickle okay (at least for known/available passed functions such as F.interpolate, torch.permute and so on)

My idea was that the module construction-time arguments are considered defaults. And then the call-time arguments can add new arguments or override the args passed at construction time. The main use-case I had before actually doesn't need overrides.

**Comment by albanD** [comment]
Yes I think there are quite a few details to iron out here (args and doc in particular as you mention).
We would accept such a function in `torch.nn.utils` if we can have a full design and some example usage for some basic Modules we have.

**Comment by vadimkantorov** [comment]
Ýep, I'll try to make a usage example. The thing is that there may be module-class gen time, module-instance construction time and forward time. Maybe not all these overrides are needed, I'll try to pick a simple one (maybe allow only args/kwargs for module-instance construction time and forward time)

**Comment by albanD** [comment]
I think that if you can come up with a subset that allows to simplify a significant number of Module declarations in core, that would be enough to add this utility.
